### PR TITLE
Create safe and unsafe versions of sparse_coo_tensor

### LINKED ISSUE #5748: sparse tensor creation should raise IndexError if indices are out of bounds
I think the following statement should raise an IndexError, indicating index 2 for axis 0 is out of bounds:
```
torch.sparse.FloatTensor(torch.LongTensor([[1, 2],[0, 2]]), torch.FloatTensor([.5,.5]), torch.Size([2,3])).to_dense()
```
At the moment it just creates the tensor in the requested size, ignoring the elements that are beyond the limits. This could introduce some bugs, that are hard to track down.

### PR DESCRIPTION
**Pull request by li-roy**
Fixes #5748.

Added an unsafe version so embedding isn't slowed.

### CODE REVIEW

File: aten/src/ATen/Declarations.cwrap
Comment (SsnL) [inline aten/src/ATen/Declarations.cwrap]: please prefix this with `_`

File: aten/src/THS/generic/THSTensor.cpp
Comment (SsnL) [inline aten/src/THS/generic/THSTensor.cpp]: Can we not have this three times? Is it possible to call `newWithTensorAndSizeUnsafe` in this can `newWithTensorAndSize`?

Comment (li-roy) [reply aten/src/THS/generic/THSTensor.cpp]: It's possible, but it would look something like the old newWithTensorSize function (because the checking happens in the middle of the function, it's hard to extract). I thought that the repetition of this one line was a better choice than having the confusing code paths.

Comment (SsnL) [reply aten/src/THS/generic/THSTensor.cpp]: It is actually three lines
```
THSTensor *self = THSTensor_(new)();
THSTensor_(rawResize)(self, nDimI, nDimV, THLongTensor_data(computed_sizes));
THSTensor_(_move)(self, THLongTensor_newWithTensor(indices), THTensor_(newWithTensor)(values));
```
With a `NB`, I would prefer to not replicate this...

Comment (li-roy) [reply aten/src/THS/generic/THSTensor.cpp]: How about refactoring those three lines into another function?

Since the checks happen after `if (sizes == NULL)` and `if (THLongTensor_nDimension(indices) == 0 && THTensor_(nDimension)(values) == 0)`, and those are required in both the safe and unsafe versions, it doesn't feel good to call the unsafe version from the safe one.

Comment (SsnL) [reply aten/src/THS/generic/THSTensor.cpp]: A local helper function would be great :)

**Comment by ezyang** [comment]
CI isn't smart enough to do the merge with `origin/master`, even though GitHub can do it: https://ci.pytorch.org/jenkins/job/pytorch-pull-request/5169/console

You'll need to rebase and repush.

**Comment by ezyang** [comment]
Nevermind, I upgraded git on Jenkins and that fixed the problem :)

**Comment by ezyang** [comment]
arrrr but the builder's versions of git are too low too lol

File: aten/src/THCS/generic/THCSTensor.cpp
Comment (SsnL) [inline aten/src/THCS/generic/THCSTensor.cpp]: Is this helper fn used in other files? If not, please static inline it.

File: aten/src/THCS/generic/THCSTensor.cpp
Comment (SsnL) [inline aten/src/THCS/generic/THCSTensor.cpp]: Why doesn't this one do the check as well?

File: aten/src/THS/generic/THSTensor.cpp
Comment (SsnL) [inline aten/src/THS/generic/THSTensor.cpp]: Same here. Why doesn't this also do the check?

**Review by ssnl — CHANGES_REQUESTED** [review]
Please also add tests for the failure cases in `test_sparse.py`. Thanks!

File: aten/src/THCS/generic/THCSTensor.cpp
Comment (li-roy) [reply aten/src/THCS/generic/THCSTensor.cpp]: The check is only relevant when sizes are given (see the old newWithTensorAndSizes, the check is only reached when sizes != none)

Comment (SsnL) [reply aten/src/THCS/generic/THCSTensor.cpp]: Ah I see. It makes sense. Thanks for explaining.

**Comment by SsnL** [comment]
Should be good to go after adding tests. :)

**Review by ssnl — APPROVED** [review]
LGTM. Let's wait till CI finishes.

**Review by ezyang — APPROVED** [review]

**Pull request merged**
